### PR TITLE
[Error handling] Add error types and error handling.

### PR DESCRIPTION
**Pull request by smcheema**
Implement error handling, as-well as error types.

### CODE REVIEW

File: replica.go [hunk 1]
```diff
@@ -66,6 +66,9 @@ func RemoveAllTagsOfReplica() ReplicaOption {
 
 // WithDemandOfReplica will add or overwrite the amount of a target Resource the replica demands
 func WithDemandOfReplica(targetResource Resource, demandAmount int64) ReplicaOption {
+	if demandAmount < 0 {
+		panic("resourceAmount cannot be negative")
+	}
```
Comment (smcheema) [inline replica.go:71]: Lots of the same checks, could potential make the `demandAmount` a custom type and implement a `.validate()` method that we can call, but seems overkill for a three line check used 3-5 times, the cost to first cast it and change it everywhere for the custom type to be accepted is not appealing to my 4AM self.

File: allocation_error.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package allocator
+
+import "fmt"
+
+type RfGreaterThanClusterSizeError int64
+type InsufficientClusterCapacityError struct{}
+type RangeWithWaywardTagsError int64
+type InvalidModelError struct{}
+type CouldNotSolveError struct{}
```
Comment (smcheema) [inline allocation_error.go:9]: Need comments.

File: allocator_test.go [hunk 3]
```diff
@@ -63,8 +63,8 @@ func TestReplicationWithInfeasibleRF(t *testing.T) {
 		clusterState.AddReplica(int64(i), rf)
 	}
 
-	status, allocation := allocator.Solve(clusterState)
-	require.False(t, status)
+	allocation, err := allocator.Solve(clusterState)
+	require.NotNil(t, err)
```
Comment (smcheema) [inline allocator_test.go:67]: `Assert` on specific error here. Create tests for other `err` types.

File: node.go [hunk 4]
```diff
@@ -56,6 +56,9 @@ func RemoveAllTagsOfNode() NodeOption {
 
 // WithResourceOfNode will add or overwrite the amount of a target Resource the node provides
 func WithResourceOfNode(targetResource Resource, resourceAmount int64) NodeOption {
+	if resourceAmount < 0 {
+		panic("resourceAmount cannot be negative")
+	}
```
Comment (smcheema) [inline node.go:61]: Fail fast on all of these annoying `< 0` checks, can't do much with them so might as well alert the caller ASAP. One hacker-mindset way of doing this is to change all fields to be `uint`, but then we're making the API obscure and losing any concrete error checking for these, also not 100% sure how the conversion from a negative integer would look like to a `uint`, might be machine/platform/sys dependent.

File: allocator.go [hunk 5]
```diff
@@ -168,39 +171,65 @@ func (a *allocator) adhereToChurnConstraint() {
...
+			for _, n := range a.nodes {
+				if replicaTagsAreSubsetOfNodeTags(r.tags, n.tags) {
+					foundNode = true
+					break
+				}
+			}
+			if !foundNode {
+				return nil, RangeWithWaywardTagsError(r.id)
+			}
+		}
```
Comment (smcheema) [inline allocator.go:221]: Might consider caching these, pre-mature optimizations are bad but caching could actually simplify some of the code in the `adhereToNodeTags()` method.

File: LICENSE [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "[]"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright [yyyy] [name of copyright owner]
```
Comment (AlirezaDX) [inline LICENSE:189]: Seems like this part needs removing/editing

Comment (smcheema) [reply LICENSE]: Done.

File: allocator.go [hunk 7]
```diff
@@ -57,20 +60,26 @@ func (a Allocation) Print() {
...
 		// compute availability of node capacity. If not defined, assume we have just enough to
 		// allocate the entire load on EACH node. This helps keep our bounds tight, as opposed to an arbitrary number.
 		if c, ok := a.nodes[0].resources[re]; ok {
 			rawCapacity = c
 		} else {
-			for _, r := range a.replicas {
-				rawCapacity += r.demands[re]
-			}
+			rawCapacity = rawDemand
```
[the remark this answers is not in the document]
Comment (smcheema) [reply allocator.go:78]: Currently model only accommodates a cluster where all nodes hold the same capacity. Have a look at the DCM meeting notes under October 29th for some of the other assumptions/simplifications we're making.

File: allocator.go [hunk 8]
```diff
@@ -57,20 +60,26 @@ func (a Allocation) Print() {
...
+		}
+
+		if rawCapacity*int64(len(a.nodes)) < rawDemand {
+			return InsufficientClusterCapacityError{}
```
[the remark this answers is not in the document]
Comment (smcheema) [reply allocator.go:82]: That's a good catch, there is a small bug in this function where I'm not accounting for `rf` inside the demand computation, let me fix it quickly. 
RE option-based-assertions, that's the idea currently, do you mean something like : 
```
if some_opt_enabled {
     validate_opt()
     act_on_opt()
}
```
I had this initially, but a lot of the validation code ends up being logic we duplicate whenever we're implementing the `opt` itself. Addition of the constraints themselves are pretty much a zero-cost operation, as long as we're erroring out before we can `.Solve()`, we don't pay any costs in term of compute time.

If I'm misunderstanding your comment can you type out some pseudocode as an example?

**Pull request merged**